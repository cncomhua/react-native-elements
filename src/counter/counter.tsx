import React, { useState } from 'react';
import { Button } from '../buttons/Button';
import { TextElement as Text } from '../text/Text';
import { normalizeText, RneFunctionComponent } from '../helpers';
import {
  View,
  StyleSheet,
  StyleProp,
  ViewStyle,
  TextStyle,
} from 'react-native';
import { withTheme } from '../config';

export type CounterProps = {
  initialValue?: number;
  min?: number;
  max?: number;
  buttonStyle?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  buttonTitleStyle?: StyleProp<TextStyle>;
  onValueChange: (number) => void;
};

interface CounterInterface extends RneFunctionComponent<CounterProps> {}

const Counter: CounterInterface = ({
  initialValue,
  min = 0,
  max = 5,
  buttonStyle,
  textStyle,
  buttonTitleStyle,
  onValueChange,
}) => {
  const [countValue, setCountValue] = useState(initialValue);

  const incrementHandler = () => {
    if (countValue < max) {
      const newValue = countValue + 1;
      setCountValue(newValue);
      onValueChange(newValue);
    }
  };

  const decrementHandler = () => {
    if (countValue > min) {
      const newValue = countValue - 1;
      setCountValue(newValue);
      onValueChange(newValue);
    }
  };

  return (
    <View style={styles.counter}>
      <Button
        testID="counterIncButton"
        onPress={decrementHandler}
        title="-"
        titleStyle={StyleSheet.flatten([styles.buttonTitle, buttonTitleStyle])}
        buttonStyle={StyleSheet.flatten([styles.button, buttonStyle])}
      />
      <Text
        style={StyleSheet.flatten([{ fontSize: normalizeText(14) }, textStyle])}
      >
        {countValue}
      </Text>
      <Button
        testID="counterDecButton"
        onPress={incrementHandler}
        title="+"
        titleStyle={StyleSheet.flatten([styles.buttonTitle, buttonTitleStyle])}
        buttonStyle={StyleSheet.flatten([styles.button, buttonStyle])}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  counter: {
    display: 'flex',
    flexDirection: 'row',
    width: 140,
    justifyContent: 'space-evenly',
    alignItems: 'center',
  },
  button: {
    borderRadius: 50,
    minWidth: 40,
  },
  buttonTitle: {
    fontSize: normalizeText(16),
    fontWeight: 'bold',
  },
});

export { Counter };
export default withTheme(Counter, 'Counter');
