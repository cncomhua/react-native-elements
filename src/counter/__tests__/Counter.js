import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Counter } from '../counter';

describe('Counter Component', () => {
  it('should render without issues', () => {
    const onValueChange = jest.fn();
    const component = shallow(
      <Counter initialValue={0} onValueChange={onValueChange} />
    );
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should have default onPress event of increment button', () => {
    const onValueChange = jest.fn();
    const wrapper = shallow(
      <Counter initialValue={0} onValueChange={onValueChange} />
    );
    wrapper.find({ testID: 'counterIncButton' }).at(0).simulate('press');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should have default onPress event of decrement button', () => {
    const onValueChange = jest.fn();
    const wrapper = shallow(
      <Counter initialValue={2} onValueChange={onValueChange} />
    );
    wrapper.find({ testID: 'counterDecButton' }).at(0).simulate('press');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should have default onPress event of decrement button and max less than value', () => {
    const onValueChange = jest.fn();
    const wrapper = shallow(
      <Counter initialValue={2} max={1} onValueChange={onValueChange} />
    );
    wrapper.find({ testID: 'counterIncButton' }).at(0).simulate('press');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should have default onPress event of decrement button and min greater than value', () => {
    const onValueChange = jest.fn();
    const wrapper = shallow(
      <Counter initialValue={1} min={2} onValueChange={onValueChange} />
    );
    wrapper.find({ testID: 'counterDecButton' }).at(0).simulate('press');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render with optional styles', () => {
    const onValueChange = jest.fn();
    const component = shallow(
      <Counter
        initialValue={0}
        onValueChange={onValueChange}
        buttonStyle={{ backgroundColor: 'black' }}
        textStyle={{ fontSize: 20 }}
        buttonTitleStyle={{ fontSize: 20 }}
      />
    );
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render without onValueChange', () => {
    const component = shallow(<Counter initialValue={0} />);
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with min max values', () => {
    const onValueChange = jest.fn();
    const component = shallow(
      <Counter initialValue={0} onValueChange={onValueChange} min={0} max={5} />
    );
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render without min max values', () => {
    const onValueChange = jest.fn();
    const component = shallow(
      <Counter initialValue={0} onValueChange={onValueChange} />
    );
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });
});
