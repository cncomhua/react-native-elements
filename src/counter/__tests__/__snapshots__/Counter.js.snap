// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component should have default onPress event of decrement button 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "space-evenly",
      "width": 140,
    }
  }
>
  <Button
    buttonStyle={
      Object {
        "borderRadius": 50,
        "minWidth": 40,
      }
    }
    onPress={[Function]}
    testID="counterIncButton"
    title="-"
    titleStyle={
      Object {
        "fontSize": 20.57142857142857,
        "fontWeight": "bold",
      }
    }
  />
  <TextElement
    h1={false}
    h1Style={Object {}}
    h2={false}
    h2Style={Object {}}
    h3={false}
    h3Style={Object {}}
    h4={false}
    h4Style={Object {}}
    style={
      Object {
        "fontSize": 18,
      }
    }
  >
    3
  </TextElement>
  <Button
    buttonStyle={
      Object {
        "borderRadius": 50,
        "minWidth": 40,
      }
    }
    onPress={[Function]}
    testID="counterDecButton"
    title="+"
    titleStyle={
      Object {
        "fontSize": 20.57142857142857,
        "fontWeight": "bold",
      }
    }
  />
</View>
`;

exports[`Counter Component should have default onPress event of increment button 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "space-evenly",
      "width": 140,
    }
  }
>
  <Button
    buttonStyle={
      Object {
        "borderRadius": 50,
        "minWidth": 40,
      }
    }
    onPress={[Function]}
    testID="counterIncButton"
    title="-"
    titleStyle={
      Object {
        "fontSize": 20.57142857142857,
        "fontWeight": "bold",
      }
    }
  />
  <TextElement
    h1={false}
    h1Style={Object {}}
    h2={false}
    h2Style={Object {}}
    h3={false}
    h3Style={Object {}}
    h4={false}
    h4Style={Object {}}
    style={
      Object {
        "fontSize": 18,
      }
    }
  >
    0
  </TextElement>
  <Button
    buttonStyle={
      Object {
        "borderRadius": 50,
        "minWidth": 40,
      }
    }
    onPress={[Function]}
    testID="counterDecButton"
    title="+"
    titleStyle={
      Object {
        "fontSize": 20.57142857142857,
        "fontWeight": "bold",
      }
    }
  />
</View>
`;

exports[`Counter Component should render with min max values 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "space-evenly",
      "width": 140,
    }
  }
>
  <Button
    buttonStyle={
      Object {
        "borderRadius": 50,
        "minWidth": 40,
      }
    }
    onPress={[Function]}
    testID="counterIncButton"
    title="-"
    titleStyle={
      Object {
        "fontSize": 20.57142857142857,
        "fontWeight": "bold",
      }
    }
  />
  <TextElement
    h1={false}
    h1Style={Object {}}
    h2={false}
    h2Style={Object {}}
    h3={false}
    h3Style={Object {}}
    h4={false}
    h4Style={Object {}}
    style={
      Object {
        "fontSize": 18,
      }
    }
  >
    0
  </TextElement>
  <Button
    buttonStyle={
      Object {
        "borderRadius": 50,
        "minWidth": 40,
      }
    }
    onPress={[Function]}
    testID="counterDecButton"
    title="+"
    titleStyle={
      Object {
        "fontSize": 20.57142857142857,
        "fontWeight": "bold",
      }
    }
  />
</View>
`;

exports[`Counter Component should render with optional styles 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "space-evenly",
      "width": 140,
    }
  }
>
  <Button
    buttonStyle={
      Object {
        "backgroundColor": "black",
        "borderRadius": 50,
        "minWidth": 40,
      }
    }
    onPress={[Function]}
    testID="counterIncButton"
    title="-"
    titleStyle={
      Object {
        "fontSize": 20,
        "fontWeight": "bold",
      }
    }
  />
  <TextElement
    h1={false}
    h1Style={Object {}}
    h2={false}
    h2Style={Object {}}
    h3={false}
    h3Style={Object {}}
    h4={false}
    h4Style={Object {}}
    style={
      Object {
        "fontSize": 20,
      }
    }
  >
    0
  </TextElement>
  <Button
    buttonStyle={
      Object {
        "backgroundColor": "black",
        "borderRadius": 50,
        "minWidth": 40,
      }
    }
    onPress={[Function]}
    testID="counterDecButton"
    title="+"
    titleStyle={
      Object {
        "fontSize": 20,
        "fontWeight": "bold",
      }
    }
  />
</View>
`;

exports[`Counter Component should render without issues 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "space-evenly",
      "width": 140,
    }
  }
>
  <Button
    buttonStyle={
      Object {
        "borderRadius": 50,
        "minWidth": 40,
      }
    }
    onPress={[Function]}
    testID="counterIncButton"
    title="-"
    titleStyle={
      Object {
        "fontSize": 20.57142857142857,
        "fontWeight": "bold",
      }
    }
  />
  <TextElement
    h1={false}
    h1Style={Object {}}
    h2={false}
    h2Style={Object {}}
    h3={false}
    h3Style={Object {}}
    h4={false}
    h4Style={Object {}}
    style={
      Object {
        "fontSize": 18,
      }
    }
  >
    0
  </TextElement>
  <Button
    buttonStyle={
      Object {
        "borderRadius": 50,
        "minWidth": 40,
      }
    }
    onPress={[Function]}
    testID="counterDecButton"
    title="+"
    titleStyle={
      Object {
        "fontSize": 20.57142857142857,
        "fontWeight": "bold",
      }
    }
  />
</View>
`;

exports[`Counter Component should render without min max values 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "space-evenly",
      "width": 140,
    }
  }
>
  <Button
    buttonStyle={
      Object {
        "borderRadius": 50,
        "minWidth": 40,
      }
    }
    onPress={[Function]}
    testID="counterIncButton"
    title="-"
    titleStyle={
      Object {
        "fontSize": 20.57142857142857,
        "fontWeight": "bold",
      }
    }
  />
  <TextElement
    h1={false}
    h1Style={Object {}}
    h2={false}
    h2Style={Object {}}
    h3={false}
    h3Style={Object {}}
    h4={false}
    h4Style={Object {}}
    style={
      Object {
        "fontSize": 18,
      }
    }
  >
    0
  </TextElement>
  <Button
    buttonStyle={
      Object {
        "borderRadius": 50,
        "minWidth": 40,
      }
    }
    onPress={[Function]}
    testID="counterDecButton"
    title="+"
    titleStyle={
      Object {
        "fontSize": 20.57142857142857,
        "fontWeight": "bold",
      }
    }
  />
</View>
`;

exports[`Counter Component should render without onValueChange 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "space-evenly",
      "width": 140,
    }
  }
>
  <Button
    buttonStyle={
      Object {
        "borderRadius": 50,
        "minWidth": 40,
      }
    }
    onPress={[Function]}
    testID="counterIncButton"
    title="-"
    titleStyle={
      Object {
        "fontSize": 20.57142857142857,
        "fontWeight": "bold",
      }
    }
  />
  <TextElement
    h1={false}
    h1Style={Object {}}
    h2={false}
    h2Style={Object {}}
    h3={false}
    h3Style={Object {}}
    h4={false}
    h4Style={Object {}}
    style={
      Object {
        "fontSize": 18,
      }
    }
  >
    0
  </TextElement>
  <Button
    buttonStyle={
      Object {
        "borderRadius": 50,
        "minWidth": 40,
      }
    }
    onPress={[Function]}
    testID="counterDecButton"
    title="+"
    titleStyle={
      Object {
        "fontSize": 20.57142857142857,
        "fontWeight": "bold",
      }
    }
  />
</View>
`;
